File: frontend/src/lib/types/entry.ts
/**
 * Prediction-entry types — mirrors `backend/app/schemas/entry.py` and
 * `backend/app/models/entry.py` after the lifecycle simplification.
 *
 * Three-status model: DRAFT, SUBMITTED, WITHDRAWN. Per-phase rows still
 * exist on the backend (Phase 2 dormant for future revival), but the UI
 * surface treats every entry as a single object with one status — the
 * Phase 1 row's status.
 */

import type { PredictionPhase } from './index';

export type EntryStatus = 'draft' | 'submitted' | 'withdrawn';

export type PaymentMode = 'per_entry' | 'per_user';

export interface EntryPhase {
	phase: PredictionPhase;
	status: EntryStatus;
	ready_at: string | null;
	submitted_at: string | null;
	locked_at: string | null;
	status_reason: string | null;
}

export interface Entry {
	id: string;
	competition_id: string;
	user_id: string;
	reference: string;
	display_name: string;
	entry_number: number;
	paid: boolean;
	prize_eligible: boolean;
	is_disabled: boolean;
	disabled_reason: string | null;
	disabled_at: string | null;
	disabled_by_user_id: string | null;
	withdrawn_at: string | null;
	withdrawn_reason: string | null;
	created_at: string;
	updated_at: string;
	phases: EntryPhase[];
}

export interface EntrySettings {
	max_entries_per_user: number;
	auto_create_first_entry: boolean;
	allow_duplicate_from_existing: boolean;
	allow_user_rename: boolean;
	payment_mode: PaymentMode;
	block_unpaid_entry_submission: boolean;
	show_entry_reference_publicly: boolean;
	phase_scoped_status_enabled: boolean;
}

export interface CompletionCount {
	done: number;
	total: number;
}

export interface CompletionSummary {
	entry_id: string;
	groups: CompletionCount;
	bracket: CompletionCount;
	bonus: CompletionCount;
	/** Team picked at `stage='winner'` in the bracket, or null if not yet
	 *  predicted. Surfaced on the entries list/card view so users see
	 *  their champion at a glance without opening the wizard. */
	predicted_winner: string | null;
}

export interface EntryCreate {
	display_name?: string;
}

export interface EntryRename {
	display_name: string;
}

export interface EntryWithdraw {
	reason: string;
}

export interface EntryReinstate {
	reason: string;
}

/**
 * Compute the entry-level status the UI should display.
 *
 * Order of precedence:
 *   1. `withdrawn` — set by admin or by the system at competition start.
 *   2. The Phase 1 row's status (`draft` / `submitted`).
 *
 * `is_disabled` is an orthogonal admin overlay (chargebacks, suspicious
 * entries). It does not replace the status badge — UI can render it as
 * an additional "Suspended" pill on top of the status if desired.
 */
export function computeDisplayStatus(
	entry: Entry,
	phase: PredictionPhase
): EntryStatus {
	if (entry.withdrawn_at) return 'withdrawn';
	const row = entry.phases.find((p) => p.phase === phase);
	return row?.status ?? 'draft';
}

/**
 * True when an entry can be edited (score inputs unlocked, "Submit"
 * affordance visible). Edit is only allowed on `draft` entries that
 * haven't crossed the competition deadline.
 */
export function isPhaseEditable(entry: Entry, phase: PredictionPhase): boolean {
	const status = computeDisplayStatus(entry, phase);
	return status === 'draft';
}

/**
 * Lifecycle button visibility helpers.
 *
 * - **Submit** — visible on DRAFT entries, before the competition deadline.
 * - **Edit**   — visible on SUBMITTED entries, before the competition deadline.
 *
 * Both predicates take the competition's `phase1_deadline` so the UI can
 * compute "is the deadline still in the future?" without hitting the
 * server again. Pass `null` if no deadline is set (admin not yet
 * configured) — both buttons are then allowed.
 *
 * Admin-only actions (withdraw, reinstate) are not exposed via these
 * helpers — they're only ever rendered from the admin entries table.
 */
export function canSubmit(
	entry: Entry,
	phase: PredictionPhase,
	phase1Deadline: string | null
): boolean {
	if (entry.is_disabled || entry.withdrawn_at) return false;
	const row = entry.phases.find((p) => p.phase === phase);
	if (!row) return false;
	if (row.status !== 'draft') return false;
	return _isBeforeDeadline(phase1Deadline);
}

export function canEdit(
	entry: Entry,
	phase: PredictionPhase,
	phase1Deadline: string | null
): boolean {
	if (entry.is_disabled || entry.withdrawn_at) return false;
	const row = entry.phases.find((p) => p.phase === phase);
	if (!row) return false;
	if (row.status !== 'submitted') return false;
	return _isBeforeDeadline(phase1Deadline);
}

function _isBeforeDeadline(deadline: string | null): boolean {
	if (!deadline) return true; // no deadline configured → editing allowed.
	return new Date(deadline).getTime() > Date.now();
}

File: backend/app/schemas/entry.py
"""Request and response schemas for prediction entry endpoints."""

from __future__ import annotations

import uuid
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from app.models.entry import EntryStatus, PaymentMode
from app.models.prediction import PredictionPhase


# ---------------------------------------------------------------------------
# Request bodies
# ---------------------------------------------------------------------------
class EntryCreate(BaseModel):
    """Optional display name. If omitted, server generates `Entry {N}`."""

    display_name: str | None = Field(default=None, min_length=1, max_length=64)


class EntryRename(BaseModel):
    display_name: str = Field(min_length=1, max_length=64)


class EntryWithdraw(BaseModel):
    """Admin withdraw — reason is REQUIRED in the simplified lifecycle.

    User-initiated withdraw was removed; admins use this with a mandatory
    reason that goes into the audit log.
    """

    reason: str = Field(min_length=1, max_length=256)


class EntryReinstate(BaseModel):
    """Admin reinstate (WITHDRAWN → SUBMITTED) — reason is REQUIRED.

    Use case: user filled all predictions but forgot to click Submit before
    competition start; admin reviews evidence and reinstates.
    """

    reason: str = Field(min_length=1, max_length=256)


class EntryDisable(BaseModel):
    """Admin disable — reason is REQUIRED per the brief."""

    reason: str = Field(min_length=1, max_length=256)


class EntryPaidUpdate(BaseModel):
    paid: bool


class EntryPrizeEligibleUpdate(BaseModel):
    prize_eligible: bool


class EntrySettingsUpdate(BaseModel):
    """Partial update — only included fields are written. All optional."""

    max_entries_per_user: int | None = Field(default=None, ge=1)
    auto_create_first_entry: bool | None = None
    allow_duplicate_from_existing: bool | None = None
    allow_user_rename: bool | None = None
    allow_user_withdrawal: bool | None = None
    require_ready_before_submit: bool | None = None
    payment_mode: PaymentMode | None = None
    block_unpaid_entry_submission: bool | None = None
    show_entry_reference_publicly: bool | None = None
    phase_scoped_status_enabled: bool | None = None
    bonus_questions_required_for_ready: bool | None = None


# ---------------------------------------------------------------------------
# Response bodies
# ---------------------------------------------------------------------------
class EntryPhaseRead(BaseModel):
    """One phase row within an entry's response payload."""

    model_config = ConfigDict(from_attributes=True)

    phase: PredictionPhase
    status: EntryStatus
    ready_at: datetime | None
    submitted_at: datetime | None
    locked_at: datetime | None
    status_reason: str | None = None


class EntryRead(BaseModel):
    """Full entry view including all phase rows."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    competition_id: uuid.UUID
    user_id: uuid.UUID
    reference: str
    display_name: str
    entry_number: int
    paid: bool
    prize_eligible: bool
    is_disabled: bool
    disabled_reason: str | None
    disabled_at: datetime | None
    disabled_by_user_id: uuid.UUID | None
    withdrawn_at: datetime | None
    withdrawn_reason: str | None
    created_at: datetime
    updated_at: datetime
    phases: list[EntryPhaseRead] = []


class EntryEventRead(BaseModel):
    """One audit row from `prediction_entry_events`."""

    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    entry_id: uuid.UUID
    phase: PredictionPhase | None
    from_status: str
    to_status: str
    actor_user_id: uuid.UUID
    actor_role: str
    reason: str | None
    created_at: datetime


class EntrySettingsRead(BaseModel):
    """All 11 competition entry settings."""

    model_config = ConfigDict(from_attributes=True)

    max_entries_per_user: int
    auto_create_first_entry: bool
    allow_duplicate_from_existing: bool
    allow_user_rename: bool
    allow_user_withdrawal: bool
    require_ready_before_submit: bool
    payment_mode: PaymentMode
    block_unpaid_entry_submission: bool
    show_entry_reference_publicly: bool
    phase_scoped_status_enabled: bool
    bonus_questions_required_for_ready: bool


class Phase2OpenResponse(BaseModel):
    """Result of admin's `/phase2/open` action."""

    entries_opened: int
    entries_skipped_withdrawn: int
    entries_skipped_disabled: int
    entries_already_open: int


class AdminEntriesPage(BaseModel):
    """Paginated response from `GET /admin/entries`.

    `items` is the entries on the requested page. `total` is the count
    of rows matching the active filters before pagination — the
    frontend uses it to render "Showing X–Y of Z" and to enable / disable
    Prev / Next buttons.
    """

    items: list[EntryRead]
    total: int


class DuplicateConflict(BaseModel):
    """Raised when a user tries to submit an identical entry.

    The frontend can deep-link to the conflicting entry via reference.
    """

    detail: str
    conflict_reference: str


class CompletionCount(BaseModel):
    done: int
    total: int


class EntryCompletionSummary(BaseModel):
    entry_id: uuid.UUID
    groups: CompletionCount
    bracket: CompletionCount
    bonus: CompletionCount
    # The team the user has picked at `stage='winner'` in their bracket
    # prediction, or `None` if not yet predicted. Surfaced on the entries
    # list / card view so users can see their champion pick at a glance
    # without opening the wizard. Added 2026-06-01.
    predicted_winner: Optional[str] = None
